File: apps/nursery-web/src/lib/api-health.ts
/**
 * Lightweight API readiness probe (no secrets).
 * Uses fetch directly so auth interceptors never run.
 */
export type ApiHealthState =
  | { status: "checking" }
  | { status: "ok" }
  | { status: "down"; detail: string };

export async function probeApiHealth(apiBaseUrl: string): Promise<ApiHealthState> {
  const base = apiBaseUrl.replace(/\/$/, "");
  const url = `${base}/health/ready`;
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), 8000);
    const res = await fetch(url, {
      method: "GET",
      headers: { Accept: "application/json" },
      signal: ctrl.signal,
      cache: "no-store",
    });
    clearTimeout(t);
    if (!res.ok) {
      return {
        status: "down",
        detail: `API health returned HTTP ${res.status}. Is the Laravel server running?`,
      };
    }
    const body = (await res.json().catch(() => null)) as
      | { status?: string; database?: string }
      | null;
    if (body?.status === "ok" || body?.database === "healthy") {
      return { status: "ok" };
    }
    return {
      status: "down",
      detail: "API responded but is not ready (database/cache). Check MySQL and .env.",
    };
  } catch {
    return {
      status: "down",
      detail:
        "Cannot reach API. Start: cd apps/nursery-api && php artisan serve --host=0.0.0.0 --port=8000. Check NEXT_PUBLIC_API_BASE_URL.",
    };
  }
}

File: apps/nursery_app/lib/core/api_health.dart
import 'package:dio/dio.dart';
import 'package:nursery_app/core/config.dart';

/// Result of GET /health/ready (no auth, no secrets logged).
class ApiHealthResult {
  const ApiHealthResult.ok() : ok = true, message = null;
  const ApiHealthResult.down(this.message) : ok = false;

  final bool ok;
  final String? message;
}

Future<ApiHealthResult> probeApiHealth() async {
  try {
    final dio = Dio(
      BaseOptions(
        baseUrl: AppConfig.apiBaseUrl,
        connectTimeout: const Duration(seconds: 8),
        receiveTimeout: const Duration(seconds: 8),
        headers: {'Accept': 'application/json'},
      ),
    );
    final res = await dio.get('/health/ready');
    final data = res.data;
    if (data is Map &&
        (data['status'] == 'ok' || data['database'] == 'healthy')) {
      return const ApiHealthResult.ok();
    }
    return const ApiHealthResult.down(
      'API responded but is not ready. Check MySQL and Laravel .env.',
    );
  } catch (_) {
    return ApiHealthResult.down(
      'Cannot reach API at ${AppConfig.apiBaseUrl}. '
      'Emulator: default 10.0.2.2. Physical phone: '
      'flutter run --dart-define=API_BASE_URL=http://YOUR_LAN_IP:8000/api/v1 '
      'and php artisan serve --host=0.0.0.0',
    );
  }
}
